hw1/hw1.py

```diff
@@ -0,0 +1,257 @@
+###### Your ID ######
+# ID1: 123456789
+# ID2: 987654321
+#####################
+
+# imports
+import numpy as np
+import pandas as pd
+
+
+def preprocess(X, y):
+    """
+    Perform mean normalization on the features and true labels.
+
+    Input:
+    - X: Input data (m instances over n features).
+    - y: True labels (m instances).
+
+    Returns:
+    - X: The mean normalized inputs.
+    - y: The mean normalized labels.
+    """
+    X_mean = X.mean(axis=0)
+    X = (X - X_mean) / (X.max(axis=0) - X.min(axis=0))
```

Comment:
Should we consider division by 0?
As well, just to be clear, this calculation works on a DataFrame with multiple columns (features), right?

Reply:
`X.mean()` calculates the mean of all the elements in the array `X`. It treats the array as a flattened sequence of values and computes the average of all the elements.

On the other hand, `X.mean(axis=0)` calculates the mean along a specific axis of the array. The `axis=0` parameter specifies that the mean should be calculated along the first axis (rows) of the array. This means that the mean will be computed for each column of the array, resulting in an array of mean values.

To illustrate this, let's consider an example. Suppose we have the following 2D array `X`:
```python
X = np.array([[1, 2, 3],
              [4, 5, 6]])
```

If we calculate `X.mean()`, it will give us the mean of all the elements in the array, which is `(1 + 2 + 3 + 4 + 5 + 6) / 6 = 3.5`.

If we calculate `X.mean(axis=0)`, it will give us the mean along the first axis (rows) of the array. This means that it will compute the mean for each column. In this case, the result will be an array `[2.5, 3.5, 4.5]`, where the first element is the mean of the first column, the second element is the mean of the second column, and so on.

About the division by 0, I do think this is a valid question to ask in Piazza. Since this actually means all elements are equal in the specific column - therefore it might be better to ignore this column completely. WDYT?

Reply:
From Piazza:
```
Q: what do we expected to do if max=min and theres div-by-0 problem?
A: Again, make sure you understand the motivation behind what we learned. In this case, if a feature is constant, is that feature beneficial?
```

As you pointed, in practice we should ignore it.